### PR TITLE
Added Multi-Host TPU tutorial

### PR DESCRIPTION
**Pull request by vfdev-5**
This is a draft of multi-host tutorial, based on this gist: https://gist.github.com/vfdev-5/70f695e462443685a0922e79ce0ee899 and Chris Jones' mnist_xla.py code.

cc @melissawm

### CODE REVIEW

File: docs/source/learn/xla-multi-host-tpu.md [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+via PCI to a board of four TPU chips (blue). The TPUs chips themselves are
+connected via high-speed inter-chip interconnects (ICI). PyTorch/XLA code runs on
+each host, e.g. via ssh. The PyTorch/XLA processes on each host are aware of each other,
+allowing you to orchestrate computation across the entire pods' worth of chips.
+[Source](https://docs.jax.dev/en/latest/multi_process.html)
+```
+
+Before diving into the code and the commands to execute, let us introduce some terminology. The section below is an adapted version of [JAX multi-host tutorial](https://docs.jax.dev/en/latest/multi_process.html#terminology).
+
+We sometimes call each Python process running PyTorch/XLA computations a controller or a host, but the terms are essentially synonymous.
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]: nit:
```suggestion
We sometimes call each Python process running PyTorch/XLA computations a _controller_ or a _host_, but the terms are essentially synonymous.
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+
+<!-- We often use the terms **global** and **local** to describe process-spanning and process-local concepts in general. For example, a "local tensor" could be a torch tensor that's only visible to a single process, compared to a PyTorch/XLA "global array" is conceptually visible to all processes. -->
+
+## MNIST example
+
+In this section we will use a simple script training a model on MNIST dataset using the SPMD model of computation and run it on a Google Cloud TPU Pod. Additionally, the training code sets up the profiler to monitor the performances per host.
+
+### Google Cloud tools setup
+
+We first need to install `gcloud` CLI. The official guide can be found [here](https://cloud.google.com/sdk/docs/install), below we provide the commands for Linux/Ubuntu:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
We first need to install the `gcloud` CLI. The [official guide](https://cloud.google.com/sdk/docs/install) has full installation instructions. Below we provide the commands for Linux/Ubuntu:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+
+<details>
+
+<summary>
+How to instal gcloud CLI on Linux/Ubuntu
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
How to install the gcloud CLI on Linux/Ubuntu
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+curl -O https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-cli-linux-x86_64.tar.gz
+
+tar -xf google-cloud-cli-linux-x86_64.tar.gz && rm -R google-cloud-cli-linux-x86_64.tar.gz && mv google-cloud-sdk /tmp/google-cloud-sdk
+
+/tmp/google-cloud-sdk/install.sh
+```
+
+</details>
+
+Next, we need to run the `gcloud` configuration by choosing the project, compute zone etc:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
Next, we need to run the `gcloud` configuration command by choosing the project and compute zone:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+gcloud init
+
+# To verify the compute zone:
+gcloud config get compute/zone
+# if we use TPUs v4, the zone should be us-central2-b
+```
+
+#### (Optional) GCS Fuse tool
+
+We can optionally install `gcsfuse` tool to be able to mount Google Cloud Storage buckets. The official guide can be found [here](https://cloud.google.com/storage/docs/cloud-storage-fuse/install), below we provide the commands for Linux/Ubuntu:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
We can optionally install the `gcsfuse` tool to be able to mount Google Cloud Storage buckets. The [official guide](https://cloud.google.com/storage/docs/cloud-storage-fuse/install) has full installation instructions. Below we provide the commands for Linux/Ubuntu:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+
+<details>
+
+<summary>
+How to instal gcsfuse CLI on Linux/Ubuntu
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
How to install the gcsfuse CLI on Linux/Ubuntu
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+# Check the installed version
+# gcsfuse -v
+```
+
+</details>
+
+
+### Python code
+
+In this section we provide the python code to train a simple neural network model on MNIST dataset using SPMD computation model. We will run this code on multi-host TPU Pod, e.g. [v8-32](https://cloud.google.com/tpu/docs/v4).
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
In this section we provide the Python code you can use to train a simple neural network model on MNIST dataset using the SPMD computation model. We will run this code on a multi-host TPU Pod, e.g. [v8-32](https://cloud.google.com/tpu/docs/v4).
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+
+The dataset will be automatically downloaded on each host. For more realistic use-cases, the training data can be fetched from a GCS bucket, for example, mounted as a folder to the filesystem using gcsfuse. For more information, please check this [documentation](https://cloud.google.com/tpu/docs/storage-options).
+
+The code snippet below contains the complete training script. Please save it into `mnist_xla.py` file. We will use this file in the next section to copy to TPU VM.
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
The code snippet below contains the complete training script. Please save it into a `mnist_xla.py` file. We will use this file in the next section to copy to TPU VM.
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+    output_folder = os.path.join(output_folder, f"proc_{host_id}")
+
+    server = xp.start_server(9012)
+    xp.start_trace(output_folder)
+    train_mnist()
+    xp.stop_trace()
+
+```
+
+Let us give now details of the training script. As we will see below when we run it on multiple TPUs, we will sent over the SSH to all workers (e.g. to 8 worker if we are using v4-32) the same command to run: `python mnist_xla.py`.
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
Let's discuss the details of this training script. As we will see below when we run it on multiple TPUs, we will send the same command over SSH to all workers (e.g. to 8 workers if we are using v4-32): `python mnist_xla.py`.
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+Each worker executes in parallel the same python commands and synchronizes with other workers on PyTorch/XLA API commands, e.g. `xr.global_runtime_device_count()`.
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
Each worker executes in parallel the same Python commands and synchronizes with other workers on PyTorch/XLA API commands, e.g. `xr.global_runtime_device_count()`.
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+
+Before calling the training method `train_mnist()` we have few lines of code to start the profiler:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
Before calling the training method `train_mnist()` we start the profiler:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+and we mark parts of the training step with `xp.Trace` to create trace events, for example:
+```python
+with xp.Trace("train_step_data_prep_and_forward"):
+    optimizer.zero_grad()
+    data, target = data.to(device), target.to(device)
+    xs.mark_sharding(data, conv_mesh, ("data", "dim1", None, None))
+    output = model(data)
+```
+
+Note: writing traces files from multiple workers to a shared folder (e.g. bucket mounted with `gcsfuse` as in our example) can present certain challenges. Namely, when multiple workers try create the same folder, `gcsfuse` can raise a "Permission denied" error. To avoid these issues we force workers to create subfolders and write traces into different folders:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
Note: writing traces files from multiple workers to a shared folder (e.g. a bucket mounted with `gcsfuse` as in our example) can present certain challenges. Namely, when multiple workers try to create the same folder, `gcsfuse` can raise a "Permission denied" error. To avoid these issues we force workers to create subfolders and write traces into different folders:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+```python
+# output_folder should exist
+output_folder = os.path.join(output_folder, f"proc_{host_id}")
+```
+
+We enable SPMD computation model with `xr.use_spmd()` and a sharded tensors will be split between the workers according to the defined meshes:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
We enable SPMD computation model with `xr.use_spmd()`, and sharded tensors will be split between the workers according to the defined meshes:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+# Global meshes definition
+num_devices = xr.global_runtime_device_count()
+device_ids = torch.arange(num_devices)
+conv_mesh_shape = (num_devices // 2, 2, 1, 1)
+conv_mesh = xs.Mesh(device_ids, conv_mesh_shape, ("data", "dim1", "dim2", "dim3"))
+linear_mesh_shape = (num_devices // 2, 2)
+linear_mesh = xs.Mesh(device_ids, linear_mesh_shape, ("data", "model"))
+```
+
+The dataloader in this script is defined using `torch.utils.data.DataLoader` and the batches in the training step with shape `(B, 1, H, W)` are sharded across the batch dimension `B` on `D1 = num_devices // 2` devices. In other words each worker keeps a local shard of shape `(B // D1, 1, H, W)` of the loaded data to form a global tensor of the training data batch of the shape `(B, 1, H, W)`. The following code is responsible for this sharding:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
The dataloader in this script is defined using `torch.utils.data.DataLoader` and the batches in the training step with shape `(B, 1, H, W)` are sharded across the batch dimension `B` on `D1 = num_devices // 2` devices. In other words, each worker keeps a local shard of shape `(B // D1, 1, H, W)` of the loaded data to form a global tensor of the training data batch of shape `(B, 1, H, W)`. The following code is responsible for this sharding:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+```python
+# Train step
+data, target = data.to(device), target.to(device)
+xs.mark_sharding(data, conv_mesh, ("data", "dim1", None, None))
+```
+The model contains convolutional and fully-connected layers and we also shard them. Convolution's weights of shape `(OutFeat, InFeat, K, K)` are sharded such that a single worker has a local shard of shape `(OutFeat // D1, InFeat, K, K)`.
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
The model contains convolutional and fully-connected layers and we also shard them. The convolution's weights of shape `(OutFeat, InFeat, K, K)` are sharded such that a single worker has a local shard of shape `(OutFeat // D1, InFeat, K, K)`.
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+The fully-connected layer `fc1` is sharded similarly to the convolutions and `fc2` are sharded over two dimensions as following `(OutFeat // D1, InFeat // 2)`. In this script the model's sharding is defined directly inside the constructor method:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
The fully-connected layer `fc1` is sharded similarly to the convolutions and `fc2` are sharded over two dimensions as following `(OutFeat // D1, InFeat // 2)`. In this script, the model's sharding is defined directly inside the constructor method:
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+
+# Create a spot TPU VM:
+gcloud compute tpus tpu-vm create $TPU_NAME --spot \
+    --accelerator-type=$ACCELERATOR_TYPE \
+    --version=$RUNTIME_VERSION \
+    --metadata-from-file=startup-script=example_startup.sh
+
+```
+
+Once the TPU VM is ready, we copy previously created `mnist_xla.py` file to all
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
Once the TPU VM is ready, we copy the previously created `mnist_xla.py` file to all
```

File: docs/source/learn/xla-multi-host-tpu.md [hunk 18]
```diff
@@ -0,0 +1,464 @@
...
+3 Train Epoch: 2 [51200/60000 (85%)]	Loss: 0.165178
+0 Train Epoch: 2 [51200/60000 (85%)]	Loss: 0.165178
+1 Train Epoch: 2 [51200/60000 (85%)]	Loss: 0.165178
+1 Training finished!
+3 Training finished!
+2 Training finished!
+0 Training finished!
+```
+
+#### Profiler logs in TensorBoard
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md:413]: I believe we want to use XProf instead of TensorBoard, but we should confirm.

File: docs/source/learn/xla-multi-host-tpu.md [hunk 19]
```diff
@@ -0,0 +1,464 @@
...
+```
+
+- Start TensorBoard:
+```bash
+tensorboard --logdir=remote_logs/ --port 7007
+```
+
+#### Troubleshooting
+
+We can execute command on all workers together or on a single worker:
```
Comment (melissawm) [inline docs/source/learn/xla-multi-host-tpu.md]:
```suggestion
We can execute commands on all workers together or on a single worker:
```

**Review by melissawm — COMMENTED** [review]
Thank you @vfdev-5 ! A few very straightforward comments and one question (should we use TensorBoard or XProf for profiling?0

**Comment by melissawm** [comment]
Hello @pgmoka @bhavya01 - would you mind taking a look for correctness and scope of this tutorial? If you are happy with the general idea, we can remove this from draft and address any other feedback. Thank you!

**Comment by melissawm** [comment]
Hi folks - gentle ping. If you have any feedback, we're happy to address. Thanks!

**Comment by melissawm** [comment]
Hi all - is this something you are still interested in? I'm happy to help bring it over the finish line if so. Thanks!

**Comment by zhanyong-wan** [comment]
@bhavya01 , could you take a look and recommend whether we should proceed with this? Thanks!

**Comment by melissawm** [comment]
Hi folks, is this still relevant or should we close? Thanks!

**Comment by bhavya01** [comment]
This is not relevant now as Pytorch/XLA is not actively maintained. I will close this PR.
